core/checks.go

```diff
@@ -253,3 +195,210 @@ func (c *Check) Query() []dagql.Selector {
+			telemetry.Boundary(),
+			// We're only using this span as a log encapsulation boundary; show
+			// its child spans inline.
+			telemetry.Passthrough(),
+		)
+		defer telemetry.EndWithCause(span, &rerr)
+		return dag.Select(ctx, dag.Root(), &checkParent, selectPath[:len(selectPath)-1]...)
+	})(ctx); err != nil {
+		return err
+	}
+
+	// Call the check
+	var status dagql.AnyResult
+	if err := dag.Select(dagql.WithRepeatedTelemetry(ctx), checkParent, &status, selectPath[len(selectPath)-1]); err != nil {
+		return err
+	}
+
+	if obj, ok := dagql.UnwrapAs[dagql.AnyObjectResult](status); ok {
+		// If the check returns a syncable type, sync it
+		if syncField, has := obj.ObjectType().FieldSpec("sync", dag.View); has {
+			if !syncField.Args.HasRequired(dag.View) {
+				if err := dag.Select(
+					dagql.WithRepeatedTelemetry(ctx),
+					obj,
+					&status,
+					dagql.Selector{Field: "sync"},
+				); err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func (c *Check) tryScaleOut(ctx context.Context) (_ bool, rerr error) {
+	q, err := CurrentQuery(ctx)
+	if err != nil {
+		return true, err
+	}
+
+	cloudEngineClient, useCloudEngine, err := q.CloudEngineClient(ctx,
+		c.Module.Source.Value.Self().AsString(),
+		// FIXME: we're saying the "function" is the check and no execCmd,
+		// which works with cloud but is weird
+		c.Name(),
+		nil,
+	)
+	if err != nil {
+		return true, fmt.Errorf("engine-to-engine connect: %w", err)
+	}
+	if !useCloudEngine {
+		return false, nil
+	}
+	defer func() {
+		rerr = errors.Join(rerr, cloudEngineClient.Close())
+	}()
+
+	query := cloudEngineClient.Dagger().QueryBuilder()
+
+	switch c.Module.Source.Value.Self().Kind {
+	case ModuleSourceKindLocal:
+		query = query.Select("moduleSource").
+			Arg("refString", filepath.Join(
+				c.Module.Source.Value.Self().Local.ContextDirectoryPath,
+				c.Module.Source.Value.Self().SourceRootSubpath,
+			))
+	case ModuleSourceKindGit:
+		query = query.Select("moduleSource").
+			Arg("refString", c.Module.Source.Value.Self().AsString()).
+			Arg("refPin", c.Module.Source.Value.Self().Git.Commit).
+			Arg("requireKind", c.Module.Source.Value.Self().Kind)
+	case ModuleSourceKindDir:
+		// FIXME: whether this actually works or not depends on whether the dir is reproducible. For simplicity,
+		// we just assume it is and will error out later if not. Would be better to explicitly check though.
+		dirID := c.Module.Source.Value.Self().DirSrc.OriginalContextDir.ID()
+		dirIDEnc, err := dirID.Encode()
+		if err != nil {
+			return true, fmt.Errorf("encode dir ID: %w", err)
+		}
+		query = query.Select("loadDirectoryFromID").
+			Arg("id", dirIDEnc)
+		query = query.Select("asModuleSource").
+			Arg("sourceRootPath", c.Module.Source.Value.Self().DirSrc.OriginalSourceRootSubpath)
+	}
+	query = query.Select("asModule")
+
+	query = query.Select("check").
+		Arg("name", c.Name())
+
+	query = query.Select("run")
+
+	query = query.SelectMultiple("completed", "passed")
```

Comment:
So essentially we create a new client on a remote engine and call the module on that engine?
Can this work recursively for sub checks?

Reply:
Yeah recursive calls actually do work, but I don't want to enable it until we have a need for it and we add detection that prevents infinite recursion, which would be a potentially expensive mistake 💸